src/server/Privilege.ts

```diff
@@ -0,0 +1,60 @@
+import { PatternDecoder, territoryPatterns } from "../core/Cosmetics";
+import { Cosmetic } from "../core/CosmeticSchemas";
+
+const patternData = territoryPatterns;
+
+export class PrivilegeChecker {
+  private patternData: Cosmetic;
+
+  constructor(patternData: Cosmetic) {
+    this.patternData = patternData;
+  }
+
+  isPatternAllowed(
+    base64: string,
+    roles: readonly string[] | undefined,
+    flares: readonly string[] | undefined,
+  ): true | "restricted" | "unlisted" | "invalid" {
+    const roleList = roles ?? [];
+    const flareList = flares ?? [];
+
+    const found = Object.entries(this.patternData.pattern).find(
+      ([, entry]) => entry.pattern === base64,
+    );
```

Comment:
The `Object.entries()` call represents a substantial allocation of many data objects on every call to this function. This allocation is not necessary, as you can directly iterate the original object for significant performance savings:
```suggestion
    let found: [string, PatternEntry] | undefined;
    for (const key in this.patternData.pattern) {
      const entry = this.patternData.pattern[key];
      if (entry.pattern === base64) {
        found = [key, entry];
        break;
      }
    }
```